File: app/src/main/java/ru/nsu/bobrofon/easysshfs/mountpointList/mountpoint/FileUtil.java
package ru.nsu.bobrofon.easysshfs.mountpointList.mountpoint;

import java.io.File;
import java.lang.reflect.Array;
import java.lang.reflect.Method;

import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.content.ContentUris;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.os.Build;
import android.os.Environment;
import android.os.storage.StorageManager;
import android.provider.DocumentsContract;
import android.provider.MediaStore;
import android.support.annotation.Nullable;

@SuppressLint("NewApi")
public final class FileUtil {

	private static final String PRIMARY_VOLUME_NAME = "primary";

	@Nullable
	public static String getFullPathFromTreeUri(@Nullable final Uri treeUri, Context con) {
		if (treeUri == null) {
			return null;
		}
		String volumePath = FileUtil.getVolumePath(FileUtil.getVolumeIdFromTreeUri(treeUri),con);
		if (volumePath == null) {
			return File.separator;
		}
		if (volumePath.endsWith(File.separator)) {
			volumePath = volumePath.substring(0, volumePath.length() - 1);
		}

		String documentPath = FileUtil.getDocumentPathFromTreeUri(treeUri);
		if (documentPath.endsWith(File.separator)) {
			documentPath = documentPath.substring(0, documentPath.length() - 1);
		}

		if (documentPath.length() > 0) {
			if (documentPath.startsWith(File.separator)) {
				return volumePath + documentPath;
			}
			else {
				return volumePath + File.separator + documentPath;
			}
		}
		else {
			return volumePath;
		}
	}


	private static String getVolumePath(final String volumeId, Context con) {
		if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
			return null;
		}

		try {
			StorageManager mStorageManager =
				(StorageManager) con.getSystemService(Context.STORAGE_SERVICE);

			Class<?> storageVolumeClazz = Class.forName("android.os.storage.StorageVolume");

			Method getVolumeList = mStorageManager.getClass().getMethod("getVolumeList");
			Method getUuid = storageVolumeClazz.getMethod("getUuid");
			Method getPath = storageVolumeClazz.getMethod("getPath");
			Method isPrimary = storageVolumeClazz.getMethod("isPrimary");
			Object result = getVolumeList.invoke(mStorageManager);

			final int length = Array.getLength(result);
			for (int i = 0; i < length; i++) {
				Object storageVolumeElement = Array.get(result, i);
				String uuid = (String) getUuid.invoke(storageVolumeElement);
				Boolean primary = (Boolean) isPrimary.invoke(storageVolumeElement);

				// primary volume?
				if (primary && PRIMARY_VOLUME_NAME.equals(volumeId)) {
					return (String) getPath.invoke(storageVolumeElement);
				}

				// other volumes?
				if (uuid != null) {
					if (uuid.equals(volumeId)) {
						return (String) getPath.invoke(storageVolumeElement);
					}
				}
			}

			// not found.
			return null;
		}
		catch (Exception ex) {
			return null;
		}
	}

	@TargetApi(Build.VERSION_CODES.LOLLIPOP)
	private static String getVolumeIdFromTreeUri(final Uri treeUri) {
		final String docId = DocumentsContract.getTreeDocumentId(treeUri);
		final String[] split = docId.split(":");

		if (split.length > 0) {
			return split[0];
		}
		else {
			return null;
		}
	}


	@TargetApi(Build.VERSION_CODES.LOLLIPOP)
	private static String getDocumentPathFromTreeUri(final Uri treeUri) {
		final String docId = DocumentsContract.getTreeDocumentId(treeUri);
		final String[] split = docId.split(":");
		if ((split.length >= 2) && (split[1] != null)) {
			return split[1];
		}
		else {
			return File.separator;
		}
	}

	public static String getPath(final Context context, final Uri uri) {

		// DocumentProvider
		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT && DocumentsContract.isDocumentUri(context, uri)) {

			if (isExternalStorageDocument(uri)) {// ExternalStorageProvider
				final String docId = DocumentsContract.getDocumentId(uri);
				final String[] split = docId.split(":");
				final String type = split[0];
				String storageDefinition;


				if("primary".equalsIgnoreCase(type)){

					return Environment.getExternalStorageDirectory() + "/" + split[1];

				} else {

					if(Environment.isExternalStorageRemovable()){
						storageDefinition = "EXTERNAL_STORAGE";

					} else{
						storageDefinition = "SECONDARY_STORAGE";
					}

					return System.getenv(storageDefinition) + "/" + split[1];
				}

			} else if (isDownloadsDocument(uri)) {// DownloadsProvider

				final String id = DocumentsContract.getDocumentId(uri);
				final Uri contentUri = ContentUris.withAppendedId(
					Uri.parse("content://downloads/public_downloads"), Long.valueOf(id));

				return getDataColumn(context, contentUri, null, null);

			} else if (isMediaDocument(uri)) {// MediaProvider
				final String docId = DocumentsContract.getDocumentId(uri);
				final String[] split = docId.split(":");
				final String type = split[0];

				Uri contentUri = null;
				if ("image".equals(type)) {
					contentUri = MediaStore.Images.Media.EXTERNAL_CONTENT_URI;
				} else if ("video".equals(type)) {
					contentUri = MediaStore.Video.Media.EXTERNAL_CONTENT_URI;
				} else if ("audio".equals(type)) {
					contentUri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI;
				}

				final String selection = "_id=?";
				final String[] selectionArgs = new String[]{
					split[1]
				};

				return getDataColumn(context, contentUri, selection, selectionArgs);
			}

		} else if ("content".equalsIgnoreCase(uri.getScheme())) {// MediaStore (and general)

			// Return the remote address
			if (isGooglePhotosUri(uri))
				return uri.getLastPathSegment();

			return getDataColumn(context, uri, null, null);

		} else if ("file".equalsIgnoreCase(uri.getScheme())) {// File
			return uri.getPath();
		}

		return null;
	}

	private static String getDataColumn(Context context, Uri uri, String selection, String[] selectionArgs) {

		Cursor cursor = null;
		final String column = "_data";
		final String[] projection = {
			column
		};

		try {
			cursor = context.getContentResolver().query(uri, projection, selection, selectionArgs, null);
			if (cursor != null && cursor.moveToFirst()) {
				final int column_index = cursor.getColumnIndexOrThrow(column);
				return cursor.getString(column_index);
			}
		} finally {
			if (cursor != null)
				cursor.close();
		}
		return null;
	}


	private static boolean isExternalStorageDocument(Uri uri) {
		return "com.android.externalstorage.documents".equals(uri.getAuthority());
	}


	private static boolean isDownloadsDocument(Uri uri) {
		return "com.android.providers.downloads.documents".equals(uri.getAuthority());
	}

	private static boolean isMediaDocument(Uri uri) {
		return "com.android.providers.media.documents".equals(uri.getAuthority());
	}

	private static boolean isGooglePhotosUri(Uri uri) {
		return "com.google.android.apps.photos.content".equals(uri.getAuthority());
	}
}

File: app/src/main/java/ru/nsu/bobrofon/easysshfs/mountpointlist/mountpoint/FileUtil.kt
// SPDX-License-Identifier: MIT
package ru.nsu.bobrofon.easysshfs.mountpointlist.mountpoint

import android.content.ContentUris
import android.content.Context
import android.database.Cursor
import android.net.Uri
import android.os.Build
import android.os.Environment
import android.os.storage.StorageManager
import android.provider.DocumentsContract
import android.provider.MediaStore
import androidx.core.net.toUri
import java.io.File
import java.lang.reflect.Array as JArray

// Some strange code from stackoverflow
object FileUtil {
    fun getFullPathFromTreeUri(treeUri: Uri, con: Context): String {
        var (volumeId, documentPath) = splitTreeUri(treeUri)
        var volumePath = volumeId?.let { getVolumePath(volumeId, con) } ?: return File.separator

        volumePath = volumePath.trimEnd(File.separatorChar)
        documentPath = documentPath.trim(File.separatorChar)

        return if (documentPath.isNotEmpty()) volumePath + File.separator + documentPath else volumePath
    }

    private fun getVolumePath(volumeId: String, context: Context): String? {
        when {
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.R -> {
                val storageManager =
                    context.getSystemService(Context.STORAGE_SERVICE) as StorageManager

                for (storageVolume in storageManager.storageVolumes) {
                    val directory = storageVolume.directory
                    if (storageVolume.isPrimary && PRIMARY_VOLUME_NAME == volumeId) {
                        return directory?.absolutePath
                    } else if (storageVolume.uuid != null) {
                        if (storageVolume.uuid.equals(volumeId)) {
                            return directory?.absolutePath
                        }
                    }
                }

                // not found.
                return null
            }

            else -> {
                try {
                    val mStorageManager =
                        context.getSystemService(Context.STORAGE_SERVICE) as StorageManager

                    val storageVolumeClazz = Class.forName("android.os.storage.StorageVolume")

                    val getVolumeList = mStorageManager.javaClass.getMethod("getVolumeList")
                    val getUuid = storageVolumeClazz.getMethod("getUuid")
                    val getPath = storageVolumeClazz.getMethod("getPath")
                    val isPrimary = storageVolumeClazz.getMethod("isPrimary")
                    val result = getVolumeList.invoke(mStorageManager) ?: return null

                    val length = JArray.getLength(result)
                    for (i in 0 until length) {
                        val storageVolumeElement = JArray.get(result, i)
                        val uuid = getUuid.invoke(storageVolumeElement) as String?
                        val primary = isPrimary.invoke(storageVolumeElement) as Boolean

                        // primary volume
                        if (primary && volumeId == "primary") {
                            return getPath.invoke(storageVolumeElement) as String
                        }

                        // other volumes
                        if (uuid == volumeId) {
                            return getPath.invoke(storageVolumeElement) as String
                        }
                    }

                    // not found
                    return null
                } catch (_: Exception) {
                    return null
                }
            }
        }
    }

    private fun splitTreeUri(treeUri: Uri): Pair<String?, String> {
        val treeDocumentId = DocumentsContract.getTreeDocumentId(treeUri)
        val split = splitId(treeDocumentId)
        val volumeId = split.getOrNull(0)
        val path = split.getOrElse(1) { "/" }

        return Pair(volumeId, path)
    }

    private fun splitId(id: String): Array<String> =
        id.split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

    fun getPath(uri: Uri, context: Context): String? = when {
        DocumentsContract.isDocumentUri(context, uri) -> getDocumentProviderPath(context, uri)
        isMediaStoreUri(uri) -> getRemoteAddress(context, uri)
        isFileUri(uri) -> uri.path
        else -> null
    }

    private fun isMediaStoreUri(uri: Uri): Boolean = "content".equals(uri.scheme, ignoreCase = true)

    private fun isFileUri(uri: Uri): Boolean = "file".equals(uri.scheme, ignoreCase = true)

    private fun getDocumentProviderPath(context: Context, uri: Uri): String? = when {
        isExternalStorageDocument(uri) -> getExternalStorageProviderPath(uri)
        isDownloadsDocument(uri) -> getDownloadsProviderPath(context, uri)
        isMediaDocument(uri) -> getMediaProviderPath(context, uri)
        else -> null
    }

    private fun getRemoteAddress(context: Context, uri: Uri): String? =
        if (isGooglePhotosUri(uri)) uri.lastPathSegment else getDataColumn(context, uri)

    private fun isGooglePhotosUri(uri: Uri): Boolean =
        uri.authority == "com.google.android.apps.photos.content"

    private fun isExternalStorageDocument(uri: Uri): Boolean =
        uri.authority == "com.android.externalstorage.documents"

    private fun isDownloadsDocument(uri: Uri): Boolean =
        uri.authority == "com.android.providers.downloads.documents"

    private fun isMediaDocument(uri: Uri): Boolean =
        uri.authority == "com.android.providers.media.documents"

    private fun getExternalStorageProviderPath(uri: Uri): String {
        val (uriType, uriValue) = splitUri(uri)

        return if ("primary".equals(uriType, ignoreCase = true)) {
            Environment.getExternalStorageDirectory().path + "/" + uriValue
        } else {
            val storageDefinition =
                if (Environment.isExternalStorageRemovable()) "EXTERNAL_STORAGE" else "SECONDARY_STORAGE"

            val storage = System.getenv(storageDefinition) ?: ""
            "$storage/$uriValue"
        }
    }

    private fun getDownloadsProviderPath(context: Context, uri: Uri): String? {
        val id = DocumentsContract.getDocumentId(uri)
        val contentUri = try {
            ContentUris.withAppendedId(
                "content://downloads/public_downloads".toUri(), java.lang.Long.valueOf(id)
            )
        } catch (e: NumberFormatException) {
            e.printStackTrace()
            return null
        }

        return getDataColumn(context, contentUri)
    }

    private fun getMediaProviderPath(context: Context, uri: Uri): String? {
        val (uriType, uriValue) = splitUri(uri)
        val contentUri: Uri? = when (uriType) {
            "image" -> MediaStore.Images.Media.EXTERNAL_CONTENT_URI
            "video" -> MediaStore.Video.Media.EXTERNAL_CONTENT_URI
            "audio" -> MediaStore.Audio.Media.EXTERNAL_CONTENT_URI
            else -> null
        }
        val selection = "_id=?"
        val selectionArgs = arrayOf(uriValue)

        return contentUri?.let { getDataColumn(context, contentUri, selection, selectionArgs) }
    }

    private fun splitUri(uri: Uri): Pair<String, String> {
        val documentId = DocumentsContract.getDocumentId(uri)
        val splitUri = splitId(documentId)

        return Pair(splitUri[0], splitUri[1])
    }

    private fun getDataColumn(
        context: Context, uri: Uri, selection: String? = null, selectionArgs: Array<String>? = null
    ): String? {
        val projection = arrayOf(DATA_COLUMN_NAME)

        return context.contentResolver.query(uri, projection, selection, selectionArgs, null)
            .use { cursor -> cursor?.let { getDataColumn(cursor) } }
    }

    private fun getDataColumn(cursor: Cursor): String? = if (cursor.moveToFirst()) {
        val columnIndex = cursor.getColumnIndexOrThrow(DATA_COLUMN_NAME)
        cursor.getString(columnIndex)
    } else null

    private const val DATA_COLUMN_NAME = "_data"

    /**
     * The name of the primary volume (LOLLIPOP).
     */
    private const val PRIMARY_VOLUME_NAME = "primary"
}
